### PR TITLE
Auto variables declaration in functions

### PR DESCRIPTION
**Pull request by laerreal**
### v2:
- extend `TestAutoDeclaration`
- refactor `VarDeclarator`

### CODE REVIEW

File: source/function/var_declarator.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        # cache
+        variables = self.variables
+
+        name2vars = defaultdict(list)
+
+        # Detect same named variables among all.
+        for v in variables:
+            name2vars[v.name].append(v)
+
+        global_vars = set(v for v in variables if v.definer is not None)
```
Comment (BDanAnd) [inline source/function/var_declarator.py]: Maybe check ```v.declarer is not None or v.definer is not None```  as in `GlobalsCollector`?
Or ```(v.declarer or v.definer) is not None``` as on 79 line?

File: source/function/var_declarator.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+        undeclared = variables - self.declared - self.arguments - global_vars
+        undeclared_names = set(v.name for v in undeclared)
+
+        # After visitting `cur` `is` `root`.
```
Comment (BDanAnd) [inline source/function/var_declarator.py:58]: Are there too many apostrophes?

Comment (laerreal) [reply source/function/var_declarator.py:58]: quite enough

File: source/function/var_declarator.py (hunk 1 above)
Comment (laerreal) [reply source/function/var_declarator.py]: Check on 79 line is enough...

**Pull request merged**